### PR TITLE
Pluggable JSON Serializer

### LINKED ISSUE #137: Dynamic types not supported in Unity with IL2CPP scripting backend causing errors.
I have been trying to use the client with Unity 3D but I am having issues since Unity does not support dynamic types (IL2CPP scripting backend) and thus errors out. Is there any way I can still use the client without using dynamics?I am restricted by certain platforms like iOS and Magic Leap which only work with IL2CPP currently.
That being said, I have had success using the client on Windows and android devices through unity and everything works well.
Would love to explore any options to still get this library to work.
Some of the other errors that I see with the IL2CPP:

2019-10-23 20:29:53.207 21466-21480/? I/Unity: try send req
    UnityEngine.Logger:Log(LogType, Object)
    Main:.ctor()
    System.Net.Http.Headers.HttpHeaders:SetValue(String, T, Func`2)
    Main:TestRequest()
     
    (Filename: ./Runtime/Export/Debug/Debug.bindings.h Line: 48)
2019-10-23 20:29:53.257 21466-21480/? I/Unity: exceptionError getting value from 'Query' on 'GraphQL.Common.Request.GraphQLRequest'.
    UnityEngine.Logger:Log(LogType, Object)
    Main:.ctor()
    System.Net.Http.Headers.HttpHeaders:SetValue(String, T, Func`2)
    Main:TestRequest()

### PR DESCRIPTION
**Pull request by rose-a**
This moves the JSON (de-)serialization out of *GraphQL.Client* to allow replacing the underlying JSON library.

Closes #137, might render #139 obsolete.

### CODE REVIEW

File: src/GraphQL.Client.Abstractions.Websocket/GraphQL.Client.Abstractions.Websocket.csproj [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
```
Comment (sungam3r) [inline src/GraphQL.Client.Abstractions.Websocket/GraphQL.Client.Abstractions.Websocket.csproj]: Why do you use netstandard2.1? What APIs from it do you use?

**Comment by sungam3r** [comment]
Relates to graphql-dotnet/server#299 , @rose-a you can look at different optimizations made in server project while moving serializers out of main code.

File: src/GraphQL.Client.Abstractions.Websocket/GraphQL.Client.Abstractions.Websocket.csproj (hunk 1 above)
Comment (rose-a) [reply src/GraphQL.Client.Abstractions.Websocket/GraphQL.Client.Abstractions.Websocket.csproj]: I guess that's a leftover from me fighting with Visual Studio... Will be removed....

**Comment by rose-a** [comment]
Thanks for the hint... odd that we're doing the same work there on 2 sides...

**Comment by benmccallum** [comment]
> Thanks for the hint... odd that we're doing the same work there on 2 sides...

Three sides if you include the [graphql-dotnet IDocumentWriter PR](https://github.com/graphql-dotnet/graphql-dotnet/pull/1512) haha.

File: src/GraphQL.Client.Abstractions.Websocket/GraphQL.Client.Abstractions.Websocket.csproj [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
+	<Import Project="../src.props" />
+
+	<PropertyGroup>
+	<TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
```
Comment (sungam3r) [inline src/GraphQL.Client.Abstractions.Websocket/GraphQL.Client.Abstractions.Websocket.csproj]: netstandard2.1

**Comment by rose-a** [comment]
The *System.Text.Json* variant currently only works with `JsonNamingPolicy.CamelCase` (which can cause a spec incompatibility, see #59). 

For *Newtonsoft.Json* this is solved by adding the `[DataMember(Name = "name")]` attributes (which are independent of the JSON library.

I saw the solution chosen for `GraphQLRequest` in graphql-dotnet/server#299, but I'm not quite sure how this could be implemented for `GraphQLResponse` and `GraphQLError` without adding yet another generic to GraphQLResponse (to be able to replace the type of `GraphQLError` inside)...

Ideas welcome! 😉

**Comment by sungam3r** [comment]
> The System.Text.Json variant currently only works with JsonNamingPolicy.CamelCase

What does that mean? You can use any policy you want.

**Comment by rose-a** [comment]
If you don't use `JsonNamingPolicy.CamelCase`, then `Data`, `Variables`, `OperationName` will be serialized as they are (with starting capital), which is not conform with the GraphQL spec.

If you use it, then you cannot send variables containing capital fields (which would be ok with the spec)

**Comment by sungam3r** [comment]
1. You can use different policies for serialization and deserialization.
2. `JsonSerializerOptions.DictionaryKeyPolicy` or `JsonSerializerOptions.PropertyNameCaseInsensitive` may help.

**Comment by rose-a** [comment]
@sungam3r Anything that should be changed in your opinion?

**Comment by sungam3r** [comment]
Firstly, I am glad that work on this project has continued. I see a lot of changes taking place. I remember very well that until recently there were a lot of not implemented features and bugs to fix. My attitude to the current process is this: I don’t really go into the details of the changes and wait for the project to stabilize. After that, I have plans to make a code review. In my projects, I completely refused to use this package just because of the lack of support. I have my own implementation that suits me. Nevertheless, if I see that the package really matured, then most likely I want to use it again and throw out my own code. That is why I intend to make a review in the future. In the meantime, I avoid watching PRs with dozens of modified files. Honestly, when I see the number 83 on the modified files tab header, I’m even scared to click on the link.

**Comment by rose-a** [comment]
Ok, no worries :wink:

**Pull request merged**